### PR TITLE
[role] ensure role definition create & update commands accept multiple permission configurations

### LINKED ISSUE #3250: [RBAC] Role definition JSON mismatch
### Description
When creating the role definition using a definition file, the nomenclature on the commands does not line up with the nomenclature in the definition file. Using the below definition file and `list` output, the following mismatches exist. See the following table for the corresponding elements. Each is mismatched, and the `--name` parameter actually corresponds to the `roleName` element output from the command. Seems like the output of the create and list operations could be aligned with the definition file and with the output from the `Get-AzureRmRoleDefinition` cmdlet.

| Definition File | Create/List Output |
|---|---|
| Name | roleName |
| Id | name |
| full resource id (not actually in file) | id |

Role definition file:
```
{
  "Name": "Role1",
  "Id": "eba5c82d-7a1c-4a3b-9a86-82a9abcfa83e",
  "IsCustom": true,
  "Description": "Test role",
  "Actions": [
    "Microsoft.Authorization/*",
    "Microsoft.Support/*"
  ],
  "NotActions": [
  ],
  "AssignableScopes": [
    "/subscriptions/a049a5ae-8a50-4a84-9a3c-abbfa1eac3a3"
  ]
}
```

Results of `list` operation for the specific role:
```
C:\WINDOWS\system32>az role definition list --name "Role1"
[
  {
    "id": "/subscriptions/a049a5ae-8a50-4a84-9a3c-abbfa1eac3a3/providers/Microsoft.Authorization/roleDefinitions/eba5c82d-7a1c-4a3b-9a86-82a9abcfa83e",
    "name": "eba5c82d-7a1c-4a3b-9a86-82a9abcfa83e",
    "properties": {
      "assignableScopes": [
        "/subscriptions/a049a5ae-8a50-4a84-9a3c-abbfa1eac3a3"
      ],
      "description": "Test role",
      "permissions": [
        {
          "actions": [
            "Microsoft.Authorization/*",
            "Microsoft.Support/*"
          ],
          "notActions": []
        }
      ],
      "roleName": "Role1",
      "type": "CustomRole"
    },
    "type": "Microsoft.Authorization/roleDefinitions"
  }
]
```

PowerShell output:
```
PS C:\> get-azurermroledefinition -name Role1


Name             : Role1
Id               : eba5c82d-7a1c-4a3b-9a86-82a9abcfa83e
IsCustom         : True
Description      : Test role
Actions          : {Microsoft.Authorization/*, Microsoft.Support/*}
NotActions       : {}
AssignableScopes : {/subscriptions/a049a5ae-8a50-4a84-9a3c-abbfa1eac3a3}
```
---

### Environment summary

**Install Method:** How did you install the CLI? (e.g. pip, interactive script, apt-get, Docker, MSI, nightly)  
Answer here:
https://docs.microsoft.com/en-us/cli/azure/install-azure-cli#windows via pip install method.

**CLI Version:** What version of the CLI and modules are installed? (Use `az --version`)  
Answer here:
azure-cli (2.0.4)

acr (2.0.2)
acs (2.0.4)
appservice (0.1.4)
batch (2.0.2)
cloud (2.0.2)
command-modules-nspkg (2.0.0)
component (2.0.3)
configure (2.0.4)
container (0.1.4)
core (2.0.4)
documentdb (0.1.4)
feedback (2.0.2)
find (0.2.1)
iot (0.1.4)
keyvault (2.0.2)
lab (0.0.3)
monitor (0.0.3)
network (2.0.4)
nspkg (3.0.0)
profile (2.0.4)
redis (0.2.1)
resource (2.0.4)
role (2.0.3)
sql (2.0.2)
storage (2.0.4)
vm (2.0.4)

Python (Windows) 3.6.1 (v3.6.1:69c0db5, Mar 21 2017, 17:54:52) [MSC v.1900 32 bit (Intel)]

**OS Version:** What OS and version are you using?  
Answer here:
Windows 10 - creators update

**Shell Type:** What shell are you using? (e.g. bash, cmd.exe, Bash on Windows)  
Answer here:
cmd

### PR DESCRIPTION
**Pull request by yugangw-msft**
This fixes #3250, that users can feed the output from the `list` command back to the `create\update` commands.
I also made refactoring to simplify the code difference caused by SDK breaking changes in the new api-version. I thought I could make swagger spec change to have the consistency in the SDK, but looks like it is not possible.

- [x] The PR has modified HISTORY.rst describing any customer-facing, functional changes. Note that this does not include changes only to help content. (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

### CODE REVIEW

**Comment by yugangw-msft** [comment]
@bganapa, can you please review?
Please note, I also added e2e test coverage for stack profile to ensure nothing are broken. Hope it doesn't conflict with the work at your end. Moving forward, please make sure to have e2e test coverages in stack profile PRs

File: src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py [hunk 1]
```diff
@@ -36,12 +37,78 @@
 # pylint: disable=too-many-lines
 
 
+class MultiAPIAdaptor(object):
```
Comment (troydai) [inline src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py]: Can you move this to its own file to limit the size of the `custom.py`?

File: src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py [hunk 2]
```diff
@@ -36,12 +37,78 @@
...
+    # We will bridge all the code difference here caused by SDK breaking change
+    def __init__(self, cli_ctx):
+        self.old_api = cli_ctx.cloud.profile == '2017-03-09-profile'
```
Comment (troydai) [inline src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py]: There are going to be three profiles now. The word `old_api` may not be precise.

**Comment by yugangw-msft** [comment]
@troydai, feedback addressed.

File: src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py [hunk 3]
```diff
@@ -405,8 +339,9 @@ def list_role_assignment_change_logs(cmd, start_time=None, end_time=None):
 
 
 def _backfill_assignments_for_co_admins(cli_ctx, auth_client, assignee=None):
+    worker = MultiAPIAdaptor(cli_ctx)
     co_admins = auth_client.classic_administrators.list()  # known swagger bug on api-version handling
-    co_admins = [x for x in co_admins if _get_role_property(x, 'email_address')]
+    co_admins = [x for x in co_admins if x.email_address]
```
Comment (bganapa) [inline src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py]: worker.get_role_property(x, 'email_address') ?

Comment (bganapa) [reply src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py]: and for the below lines where we refer to email_address

File: src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py [hunk 4]
```diff
@@ -417,30 +352,36 @@ def _backfill_assignments_for_co_admins(cli_ctx, auth_client, assignee=None):
...
+        email = admin.email_address
         result.append({
             'id': na_text,
             'name': na_text,
+        })
+        properties = {
             'principalId': upns.get(email),
             'principalName': email,
-            'roleDefinitionName': _get_role_property(admin, 'role'),
+            'roleDefinitionName': admin.role,
```
Comment (bganapa) [inline src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py]: this one also need to go thru worker?

File: src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py (hunk 3 above)
Comment (yugangw-msft) [reply src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py]: no needed, classic-admin's api-version never changed.

File: src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py (hunk 4 above)
Comment (yugangw-msft) [reply src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py]: Same, classic-admin's api-version never changed.

File: src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py (hunk 3 above)
Comment (bganapa) [reply src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py]: Thanks

**Pull request merged**